ADR-74: Migrate Timeout Parameters to Consensus Parameters (#7503)
tendermint/tendermint: docs/architecture/adr-074-timeout-params.md

from each other. There is no reason for validators on the same network to configure
different timeout values. Proper functioning of the Tendermint consensus algorithm
relies on these parameters being uniform across validators.

The configurable values are as follows:

* `TimeoutPropose`
	* How long the consensus algorithm waits for a proposal block before issuing a prevote.
	* If no prevote arrives by `TimeoutPropose`, then the consensus algorithm will issue a nil prevote.
* `TimeoutProposeDelta`
	* How much the `TimeoutPropose` grows each round.
* `TimeoutPrevote`
	* How long the consensus algorithm waits after receiving +2/3 prevotes with
	no quorum for a value before issuing a precommit for nil.
	(See the [arXiv paper][arxiv-paper], Algorithm 1, Line 34)
* `TimeoutPrevoteDelta`
	* How much the `TimeoutPrevote` increases with each round.
* `TimeoutPrecommit`
	* How long the consensus algorithm waits after receiving +2/3 precommits that
	do not have a quorum for a value before entering the next round. 
	(See the [arXiv paper][arxiv-paper], Algorithm 1, Line 47)
* `TimeoutPrecommitDelta`
	* How much the `TimeoutPrecommit` increases with each round.

> * `TimeoutCommit`

Reviewer: So just to confirm, this timer starts after we call the `Commit` ABCI method right? Not immediately after receiving 2/3+ precommits

Author: Double checking this, I'm seeing that we calculate the `StartTime` for the next height using the `CommitTime` of the previous height.

`CommitTime` is set to the be the time that the validator received +2/3 precommits:
https://github.com/tendermint/tendermint/blob/7cdf560173dee6773b80d1c574a06489d4c394fe/internal/consensus/state.go#L1598

In effect, this means that `TimeoutCommit` starts when +2/3 precommits are received, not after the application has finished the computation for the commit. 

What tripped me up is that we set the variable after the commit completes, but it used the previously captured `CommitTime` value.

Reviewer: Okay. This seems reasonable to me. Thanks for double checking

# ADR 74: Migrate Timeout Parameters to Consensus Parameters

## Changelog

- 03-Jan-2022: Initial draft (@williambanfield)

## Status

Proposed

## Context

### Background

Tendermint's consensus timeout parameters are currently configured locally by each validator
in the validator's [config.toml][config-toml].
This means that the validators on a Tendermint network may have different timeouts
from each other. There is no reason for validators on the same network to configure
different timeout values. Proper functioning of the Tendermint consensus algorithm
relies on these parameters being uniform across validators.

The configurable values are as follows:

* `TimeoutPropose` 
	* How long the consensus algorithm waits for a proposal block before issuing a prevote.
	* If no prevote arrives by `TimeoutPropose`, then the consensus algorithm will issue a nil prevote.
* `TimeoutProposeDelta` 
	* How much the `TimeoutPropose` grows each round.
* `TimeoutPrevote` 
	* How long the consensus algorithm waits after receiving +2/3 prevotes before 

> 	issuing a precommit.

Author: I found this to be a bit confusing as well, but this is the way this works. See the [arxiv paper](https://arxiv.org/pdf/1807.04938.pdf), line 34 and line 47. Loosely: it bounds how many steps apart any two validators may be during consensus.

* `TimeoutPrevoteDelta` 
	* How much the `TimeoutPrevote` increases with each round.
* `TimeoutPrecommit` 
	* How long the consensus algorithm waits after receiving +2/3 precommits with 
	no quorum before entering the next round.
* `TimeoutPrecommitDelta` 
	* How much the `TimeoutPrecommit` increases with each round.
* `TimeoutCommit` 
	* How long the consensus algorithm waits after committing a block but before starting the new height.
	* This gives a validator a chance to receive slow precommits.
* `SkipTimeoutCommit`

> 	* Make progress as soon as the node has 100% of the precommits.

Author: I think that this is just somewhat confusingly documented, but this comment is a bit clarifying: https://github.com/tendermint/tendermint/blob/15a3110a6ac2fe769cee0c6cd2c5ad40eabb689c/internal/consensus/state.go#L1065

Additionally, see where the configured `bool` is used:
https://github.com/tendermint/tendermint/blob/15a3110a6ac2fe769cee0c6cd2c5ad40eabb689c/internal/consensus/state.go#L2108

https://github.com/tendermint/tendermint/blob/15a3110a6ac2fe769cee0c6cd2c5ad40eabb689c/internal/consensus/state.go#L2233

Notice that it's calling `HasAll`, which is a method that checks to ensure that 100% of the voting power has been seen:
https://github.com/tendermint/tendermint/blob/15a3110a6ac2fe769cee0c6cd2c5ad40eabb689c/types/vote_set.go#L428

> 	* Make progress as soon as the node has 100% of the precommits.

Reviewer: Right, thanks for the explanation. Yeah I think we totally need to better document what this means. I wonder if we can even come up with a better, more intuitive, name for the variable?

> 	* Make progress as soon as the node has 100% of the precommits.

Author: Yes, I definitely agree that the docs can be better on this. Maybe something like 'timeout-commit-enable-bypass' which makes it clear that's 'enabled' unlike 'skip' which sounds like it happens every time.

> 	* Make progress as soon as the node has 100% of the precommits.

Author: Updated in this doc.

### Overview of Proposed Change

This ADR proposes consolidating the timeout parameters and migrating them from the
node-local `config.toml` file into the network-global consensus parameters.
The consensus parameters are the same across all validators on the network.

> A sane default will be provided by Tendermint for each of these parameters and

Author: Good suggestion, I appreciate the language change 🙏

> networks will be able to update the parameters as they see fit. Local updates

Reviewer: Have you thought about what the logistics are around an upgrade? Is there any possibility that the application could set the defaults or must they take our defaults and only change them after a governance proposal is passed?

Author: > Is there any possibility that the application could set the defaults or must they take our defaults and only change them after a governance proposal is passed?

This is an important question, I'll do a bit of research to see what options may exist to allow application developers to pick values ahead of the change.

Author: Applications could bypass the governance mechanism by returning values at a specific height. The application would have to maintain the logic to add something at a certain height forever to allow validators that join the chain after the change has taken effect to receive the param change.

Author: Added a migration section :+1:

Reviewer: I guess this would also apply to PBTS so having an appropriate mechanism for application developers to set the initial values seems like a good idea. Have you thought perhaps about using `InitChain` upon upgrade to fetch the new value?

Author: I think that re-purposing that call is pretty risky. We currently document that call as only being made on chain-start and we don't provide a signal for applications to actually determine the height at which the call is being made so it's not obvious how developers could distinguish between 'chain start' and 'upgrade' 

This also would not be present for nodes that are syncing to the network. If the upgrade happened at height 100 and a node joined after the upgrade, there is not obvious way for that application to learn that a consensus param changed at height 100. This is, imo, why keeping it in EndBlock is pretty reasonable.

This does leave at least one height that must use the defaults, i.e. the first height after the upgrade where the new params exist but are not yet set. Given that the default params do not appear to be changed by popular chains makes this seem very low risk, but I definitely see it as pretty inelegant. This is a good candidate for discussion in our next dev session. I'll add it to the agenda there.

Reviewer: > it's not obvious how developers could distinguish between 'chain start' and 'upgrade'

I was thinking of the case where it's one and the same binary in which case the developers don't really need to know what height it is at.

I think you're probably right about it not being worth it, I just think that this might be a common occurrence and perhaps we develop something specifically for that use case.

Author: Yeah, this is a pretty complex thing to do. Captured thoughts here: https://github.com/tendermint/tendermint/pull/7524

to these parameters will no longer be possible; instead, the application will control
updating the parameters. Applications using the Cosmos SDK will be automatically be
able to change the values of these consensus parameters [via a governance proposal][cosmos-sdk-consensus-params].
This ADR also proposes consolidating the 8 timeout parameters into 6. These
6 parameters would be as follows:

* `TimeoutPropose`
	* Same as current `TimeoutPropose`.
* `TimeoutProposeDelta`
	* Same as current `TimeoutProposeDelta`.
* `TimeoutVote`
	* Parameter that governs how long validators wait for votes in both the prevote
	 and precommit phase of the consensus algorithm. This parameter will subsume
	 the current `TimeoutPrevote` and `TimeoutPrecommit` parameters.
* `TimeoutVoteDelta`
	* Parameter that governs how much the `TimeoutVote` will grow each successive round.
	 This parameter will subsume the current `TimeoutPrevoteDelta` and `TimeoutPrecommitDelta`
	 parameters.
* `TimeoutCommit`
	* Same as current `TimeoutCommit`.
* `SkipTimeoutCommit`

> 	* Same as current `SkipTimeoutCommit`.

Author: explanation here: https://github.com/tendermint/tendermint/pull/7503#discussion_r778336135

This change is low risk. While these values may be configured, many running chains
do not change these parameters from their default values. For example, initializing
a node on Osmosis, Terra, and the Cosmos Hub using the their `init` command produces
a `config.toml` with Tendermint's default values for these parameters.

### Why this parameter consolidation?

Reducing the number of parameters is good for UX. Fewer superfluous parameters makes
running and operating a Tendermint network less confusing. 

The Prevote and Precommit messages are both similar sizes and are also similar in

> their function in the consensus algorithm so there is no strong need for them to

Author: Good correction, fixed.

be configured separately.

The `TimeoutPropose` parameter governs how long Tendermint will wait for the proposed
block to be gossiped. Blocks are much larger than votes and therefore tend to be
gossiped much more slowly. It therefore makes sense to keep `TimeoutPropose` and

> the `TimeoutProposeDelta` as parameters separate from the vote timeouts.

Author: Yeah, it effectively allows the algorithm to 'search' for the correct timeouts so that it can adapt to changes in network conditions/topology and block size.

> the `TimeoutProposeDelta` as parameters separate from the vote timeouts.

Reviewer 2: Not that it matters for this ADR, but how does it adjust back down? Or does it reset to the base on every new block, discarding adjustments from the rounds of the previous height?

> the `TimeoutProposeDelta` as parameters separate from the vote timeouts.

Author: Resets back down to the base at each height and discards the adjustments from the previous height.
There shouldn't be any issue with changing the logic to stay at the higher value for the next round or 'search' for a large enough time in a different way.

> the `TimeoutProposeDelta` as parameters separate from the vote timeouts.

Reviewer: And if it's consistently too short then now they the network can agree to adjust it to a longer time

> `TimeoutCommit` is used by chains to ensure that the network waits for the votes from

Author: > For example, say I run a cabal of fast validators on good network, with near or over 2/3 voting power. How am I enforced against patching my validators to skip the block delay and start the next round immediately? In this sense, my validators are Byzantine—they violate the protocol—but they do so in a way that's hard to detect.

There is nothing that stops this and we do not currently generate evidence about this behavior. Here are the types of evidence we do collect:
https://github.com/tendermint/tendermint/blob/2c55eead7c71f4686711bcec198de8bd7a3d4714/proto/tendermint/types/evidence.pb.go#L76

This would be difficult evidence to collect because you would need proof that a validator collected the precommits from a height and executed the associated block _later_ than it claims to have done so but there is no trusted way currently to gather information about when a validator gathered its data. The best I believe you would be able to do would be determine if a validator issued a proposal/prevote/precommit before the time `2/3rth_precommit_issued + timeout_commit` and generate evidence to prevent egregiously bad behavior.

I agree that it would be a byzantine behavior and it would confer some degree of economic advantage to the fast validators: they would be able to freeze out the other validators as proposers and therefore collected the block rewards for proposing.

require a longer `TimeoutCommit` to allow the application to complete its work
and be prepared for the next height.

* From https://github.com/tendermint/spec/issues/359 it looks like `TimeoutCommit` really only exists for the purpose of allowing
precommits to be collected after the +2/3 has initially been seen. Not much more value to it than that. 

* allowing long timeout commit discussed here: https://github.com/tendermint/tendermint/issues/5911#issuecomment-973560381

## Decision

None

## Detailed Design

### New Consensus Parameters

A new `TimeoutParams` `message` will be added to the [params.proto file][consensus-params-proto].
This message will have the following form:

```proto
message TimeoutParams {
 google.protobuf.Duration timeout_propose = 1
       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
 google.protobuf.Duration timeout_propose_delta = 2
       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
 google.protobuf.Duration timeout_vote = 3
       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
 google.protobuf.Duration timeout_vote_delta = 4
       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];

>  google.protobuf.Duration timeout_commit = 5

Reviewer: I wonder if having the `timeout` prefix to all these variables causes it to stutter i.e. `params.timeout.timeout_propose`

Author: done.

       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
 bool skip_timeout_commit = 6;
}
```

This new `message` will then be added as a field into the [`ConsensusParams`
message][consensus-params-proto]. The same default values that are [currently
set for these parameters][current-timeout-defaults] in the local configuration
file will be used as the defaults for these new consensus parameters in the
[consensus parameter defaults][default-consensus-params].

> Validation of these new parameters will be identical to their [validation
> currently][time-param-validation].
> Namely, these parameters must be non-negative.

Reviewer 2: Suggestion to maybe clarify which parameters we mean in each case:

```suggestion
The new consensus parameters will be subject to the same
[validity rules][time-param-validation] as the current configuration values,
namely, each value must be non-negative.
```

### Removal of Old Parameters

> The old `timeout-*` parameters that are configured in the [config.toml][config-toml]

Reviewer 2: We should probably stage this somehow, even if we don't allow temporary overrides. For example:

1. In release v0.36, a node receiving a block without timeouts in the consensus parameters will fall back on the local settings (this allows node software to be upgraded without stopping the chain)
2. In release v0.36, if any non-default timeout values are set in the config, the node will log a warning when it starts up about the upcoming deprecation of those values.
3. In release v0.37, the parameters are suppressed or removed, so that starting up a node which has them in its config will print a specific error pointing to the documentation (rather than just "unknown field lol")

(This is just a sketch of how it could be done)

The other thing I wonder is how do we bootstrap these values into consensus parameters, on chains that don't already have them? For the defaults, of course, it's a noop. But if I wanted to have a longer block delay, say, can we do that as a "soft" migration?

Author: > The other thing I wonder is how do we bootstrap these values into consensus parameters, on chains that don't already have them? For the defaults, of course, it's a noop. But if I wanted to have a longer block delay, say, can we do that as a "soft" migration?

Unless I'm misunderstanding the question, this should be pretty straightforward. The values themselves are not actually included in the blocks, just a hash of the values. The hash already exists, so we just have to make sure to make the change in a way that preserves the old hashes.

will be removed completely.

### Optional: Temporary Local Overrides

The new `TimeoutParams` will be released during the `v0.36` release cycle. Accidentally
configuring these parameters too low would result could result in chains with slow
networks or low-degree network topologies to not gossip votes within the configured
timeouts and require many rounds for consensus to occur. To prevent this condition,
we could optionally include a set of `unsafe-timeout-*` parameters in the `config.toml`.

These parameters would allow node operators to quickly remedy the situation while
preparing to update the consensus parameters. This would be a temporary solution that
we would remove within the v0.37 release.

### Add New Consensus Parameters to HashedParams

Tendermint currently only verifies that a subset of the consensus parameters are

> equal across all validators. These parameters are the `BlockMaxBytes` and the `BlockMaxGas`.

Reviewer: Do we only verify that these parameters are the same, not any of the others 🤯 ?

Author: Yup :) only these ones
https://github.com/tendermint/tendermint/blob/15a3110a6ac2fe769cee0c6cd2c5ad40eabb689c/types/params.go#L185

A [hash of these parameters][hashed-params] is included in the block. Validators ensure 
their values of the parameters match by hashing their value of the parameters and 
checking that their hashed value matches the hash included in the block. 

> Including the new parameters in the consensus parameters hash will break verification

Reviewer 2: Could we finesse this by only hashing non-default values?

Reviewer 2: I think the only gotcha here is that it means "the defaults at the time of the initial upgrade". We'd have to pin those against other changes in the defaults we might make later.

Author: That's a crucial caveat. 

updated the doc to reflect this upgrade scheme.

and [spec issue 359][spec-359].

## Alternative Approaches

### Hardcode the parameters

Many Tendermint networks run on similar cloud-hosted infrastructure. Therefore,
they have similar bandwidth and machine resources. The timings for propagating votes
and blocks are likely to be reasonably similar across networks. As a result, the
timeout parameters are good candidates for being hardcoded. Hardcoding the timeouts
in Tendermint would mean entirely removing these parameters from any configuration
that could be altered by either an application or a node operator. Instead,
Tendermint would ship with a set of timeouts and all applications using Tendermint
would use this exact same set of values.

While Tendermint nodes often run with similar bandwidth and on similar cloud-hosted
machines, there are enough points of variability to make configuring
consensus timeouts meaningful. Namely, Tendermint network topologies are likely to be
very different from chain to chain. Additionally, applications may vary greatly in 
how long the `Commit` phase may take. Applications that perform more work during `Commit`
require a longer `TimeoutCommit` to allow the application to complete its work

> and be prepared for the next height.

Reviewer: Also don't forget that the size of a block and the number of validators could be vastly different

* From https://github.com/tendermint/spec/issues/359 it looks like `TimeoutCommit` really only exists for the purpose of allowing
precommits to be collected after the +2/3 has initially been seen. Not much more value to it than that.

* allowing long timeout commit discussed here: https://github.com/tendermint/tendermint/issues/5911#issuecomment-973560381

## Decision

> None

Author: was leaving til we seemed to have consensus.

## Detailed Design

### New Consensus Parameters

A new `TimeoutParams` `message` will be added to the [params.proto file][consensus-params-proto].
This message will have the following form:

```proto
message TimeoutParams {
 google.protobuf.Duration propose = 1;
 google.protobuf.Duration propose_delta = 2;
 google.protobuf.Duration vote = 3;
 google.protobuf.Duration vote_delta = 4;
 google.protobuf.Duration commit = 5;
 bool commit_timeout_bypass_enabled = 6;
}
```

This new message will be added as a field into the [`ConsensusParams`
message][consensus-params-proto]. The same default values that are [currently
set for these parameters][current-timeout-defaults] in the local configuration
file will be used as the defaults for these new consensus parameters in the
[consensus parameter defaults][default-consensus-params].

The new consensus parameters will be subject to the same
[validity rules][time-param-validation] as the current configuration values,
namely, each value must be non-negative.

### Migration

Migration to using these new parameters will occur across two releases.
During the v0.36 release, Tendermint will add these as consensus parameters.

> The old parameters config-local parameters will continue to exist as well in the

Reviewer: ```suggestion
The old local config parameters will continue to exist as well in the
```

> The old parameters config-local parameters will continue to exist as well in the

Reviewer: ```suggestion
The old local config parameters will continue to exist as well in the
```

v0.36 release. If the parameters are set as consensus parameters, the config-local
parameters will be ignored. If the consensus parameters are not set, Tendermint

> will fall back to using the values contained in the `config.toml` file.

Reviewer: Does this mean that we are allowing networks to ignore the consensus params if they want and still use every nodes local config parameters. Are we sure we want to do this?

Author: Once the network sets a non-default values in the consensus parameters, we will use those. This just gives a release during which networks can make the upgrade.

I think that this is a good way for networks to gracefully upgrade with minimal disruption. The values in the config file are strictly fallbacks in case no values are set by the chain. It maintains a partial version of the already kind of bad thing for a release so that we can more seamlessly transition to better management of these params.

During the v0.36 release cycle, if non-default values are set in the `config.toml`,
the node will log a warning alerting operators that the parameters will soon be removed

> in the upcoming release.

Reviewer: But more importantly, will these non-default values be used?

Author: They will only be used if the chain does not set them as consensus parameters.

Reviewer: But what do you mean by not setting the consensus params. Aren't they set to some reasonable default upon initialization? How would they be unset?

Author: So basically it would be as follows:
If the chain has not yet set them in EndBlock, use what is in `config.toml`.
Once the chain sets them in `EndBlock`, use those values and ignore `config.toml`.

Reviewer: Would this work with state syncing? Will we need an extra flag to tell us if it has been set or not? I think it would be better if we transition to using consensus params directly when the chain starts/restarts i.e. use the genesis values or the defaults in the case of new params in an upgraded chain (I know you've written an entire RFC on this so perhaps I should check there first)

Author: > Would this work with state syncing? Will we need an extra flag to tell us if it has been set or not?

We would not need this flag, no. We would just always treat 'value as absent' as 'value uses default'. This way historical blocks would always look to be using the defaults. Once the node that is syncing hits a height where the application changed the parameters via `EndBlock` it would then start to use the updated values. So far this is fine because all param changes we are proposing are timing related and validation of historical blocks does not consider timing at all.

	 parameters.
* `TimeoutCommit`
	* Same as current `TimeoutCommit`.

> * `TimeoutCommitBypassEnabled`

Reviewer 2: I agree with the reasoning here—but since all the other names here beginning with "Timeout" are, in fact, timeouts, what about maybe: `EnableTimeoutCommitBypass` to signal that this one is a flag?

Reviewer: `EnableTimeoutCommitBypass` doesn't sound too different to `SkipTimeoutCommit`. Maybe I'm being too fussy but I think we shouldn't refer to `TimeoutCommit` in this flag as they are really triggered at two different thresholds. TimeoutCommit starts at 2/3, this bypass is triggered once we receive 100% of the votes. So they kind of have two different starting points. `EnableProgressionAfterAllVotes`? `EnableInstantProgression`. Maybe there's no right way to name this variable and we just need to be sure that the documentation is comprehensive

Author: I think that the logic is somewhat too complex to describe in shortish variable name.

Reviewer: Okay, I only ask then that perhaps we don't use `TimeoutCommit` in the name of this variable. I think it's better if we don't try to associate these two variables together
